tests/perf-system/benchmarks/benchmarks.py

```diff
@@ -0,0 +1,513 @@
+                metric["errors"] == 0
+            ), f"Expected 0 errors, got {metric['errors']}. Full metrics: {metrics}"
+            assert (
+                metric["failures"] == 0
+            ), f"Expected 0 failures, got {metric['failures']}. Full metrics: {metrics}"
+            assert (
+                metric["retries"] == 0
+            ), f"Expected 0 retries, got {metric['retries']}. Full metrics: {metrics}"
+
+            # todo: handle case that metric is missing & handle forwarding better
+            if metric["path"] == "log/public":
+                if metric["method"] == "POST":
+                    assert (
+                        metric["calls"] >= total_posts
+                    ), f"Expected {total_posts} posts, got {metric['calls']}. Full metrics: {metrics}"
+                elif metric["method"] == "GET":
+                    assert (
+                        metric["calls"] >= total_gets
+                    ), f"Expected {total_gets} gets, got {metric['calls']}. Full metrics: {metrics}"
+                else:
+                    assert (
+                        False
+                    ), f"Unexpected method {metric['method']}. Full metrics: {metrics}"
+            else:
+                assert metric["path"] in [
+                    "api/metrics",
+                    "commit",
+                    "log/public/count",
+                ], f"Unexpected path {metric['path']}"
+
+        LOG.info("Checking node metrics")
+        node_metrics_process = subprocess.run(
+            [
+                "curl",
+                address + "/node/metrics",
+                *user_cert_args,
+            ],
+            capture_output=True,
+        )
+        node_metrics = json.loads(node_metrics_process.stdout.decode())["sessions"]
+        actual_peak_sessions = int(node_metrics["peak"])
+        expected_peak_sessions = max(ccf_node_for_piccolo.count(ccf_node_ips[id]), 1)
+        assert (
+            actual_peak_sessions == expected_peak_sessions
+        ), f"Expected {expected_peak_sessions} peak sessions, got {actual_peak_sessions}"
+
+        LOG.info("Checking commit tx term")
+        commit_tx_process = subprocess.run(
+            [
+                "curl",
+                address + "/commit",
+                *user_cert_args,
+            ],
+            capture_output=True,
+        )
+        commit_tx = json.loads(commit_tx_process.stdout.decode())["transaction_id"]
+        assert int(commit_tx.split(".")[0]) == 2, f"Commit tx is {commit_tx}"
+
+        node_ip = ccf_node_ips[id].split(":")[0]
+        LOG.info(f"Fetching workspace from node {id} at {address}")
+        sandbox_dir = experiment_dir / "workspace" / f"sandbox_{id}"
+        os.makedirs(sandbox_dir)
+        fetch_remote_dir(node_ip, ccf_dir / "workspace" / f"sandbox_{id}", sandbox_dir)
+        LOG.info("Fetching completed")
+
+    LOG.info("Shutting down CCF service")
+    ccf_server_process.send_signal(signal.SIGINT)
```

Comment:
It's possible to obtain the same result without delay by allocating a PTY, setting PR_SET_PDEATHSIG and simply closing the connection (or exiting the scope when using a context manager).